options_tester.md
Loading config file for options_tester
## Testing SHORT positions


#### TESTING POSITION 1/5 (ENTRY)
 In order to enter a SHORT position (PUT) in $(F) with goal ASSET_SET_PRICE of $15.05 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 15.04
asset_change: 0.06
asset_percentage_change: 0.4
bid: 0.36
ask: 0.39
greeks: {'delta': -0.4700047124101207, 'theta': -0.01547070416850569, 'gamma': 0.4649728827864585, 'vega': 0.010386708432297337, 'implied_violatility': 32.7683687210083}
```


#### TESTING POSITION 1/5(EXIT)
 In order to exit a SHORT position (PUT) in $(F)  with goal ASSET_SET_PRICE of $14.82 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 14.82
asset_change: -0.16
asset_percentage_change: -1.07
bid: 0.39
ask: 0.47
greeks: {'delta': -0.5325491744903564, 'theta': -0.02514211157523686, 'gamma': 0.28582373787022064, 'vega': 0.010229619412039613, 'implied_violatility': 54.07106876373291}
```


#### TESTING POSITION 2/5 (ENTRY)
 In order to enter a SHORT position (PUT) in $(F) with goal ASSET_PERCENTAGE of 0.6% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

0.6 percent of $14.98: $0.08988
```
asset_price: 15.06
asset_change: 0.08
asset_percentage_change: 0.53
bid: 0.35
ask: 0.39
greeks: {'delta': -0.4596478395639391, 'theta': -0.014549801653071676, 'gamma': 0.4921323774108526, 'vega': 0.010376615589126045, 'implied_violatility': 30.847787857055664}
```


#### TESTING POSITION 2/5(EXIT)
 In order to exit a SHORT position (PUT) in $(F)  with goal ASSET_PERCENTAGE of 0.8% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

0.8 percent of $14.98: $0.11984
```
asset_price: 14.86
asset_change: -0.12
asset_percentage_change: -0.8
bid: 0.39
ask: 0.45
greeks: {'delta': -0.525567561900108, 'theta': -0.023418893093551173, 'gamma': 0.30764145306087953, 'vega': 0.01027036005443121, 'implied_violatility': 50.16529560089111}
```


#### TESTING POSITION 3/5 (ENTRY)
 In order to enter a SHORT position (PUT) in $(F) with goal OPTION_SET_PRICE of $0.35 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 15.06
asset_change: 0.08
asset_percentage_change: 0.53
bid: 0.35
ask: 0.39
greeks: {'delta': -0.45962350266123736, 'theta': -0.014531072172559027, 'gamma': 0.4927605750510875, 'vega': 0.01037655110231925, 'implied_violatility': 30.80826997756958}
```


#### TESTING POSITION 3/5(EXIT)
 In order to exit a SHORT position (PUT) in $(F)  with goal OPTION_SET_PRICE of $0.45 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 14.86
asset_change: -0.12
asset_percentage_change: -0.8
bid: 0.39
ask: 0.45
greeks: {'delta': -0.5256030566114234, 'theta': -0.023404957724889132, 'gamma': 0.3078211009767406, 'vega': 0.010270301290477114, 'implied_violatility': 50.13573169708252}
```


#### TESTING POSITION 4/5 (ENTRY)
 In order to enter a SHORT position (PUT) in $(F) with goal OPTION_TICK of $0.05 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 15.08
asset_change: 0.1
asset_percentage_change: 0.67
bid: 0.34
ask: 0.39
greeks: {'delta': -0.44781585156900194, 'theta': -0.013589404352283996, 'gamma': 0.5232671024738367, 'vega': 0.010354390001262122, 'implied_violatility': 28.873443603515625}
```


#### TESTING POSITION 4/5(EXIT)
 In order to exit a SHORT position (PUT) in $(F)  with goal OPTION_TICK of $0.15 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 14.69
asset_change: -0.29
asset_percentage_change: -1.94
bid: 0.39
ask: 0.54
greeks: {'delta': -0.5485645394269146, 'theta': -0.030670372026068378, 'gamma': 0.2323863462959203, 'vega': 0.01009829078866279, 'implied_violatility': 66.81805849075317}
```


#### TESTING POSITION 5/5 (ENTRY)
 In order to enter a SHORT position (PUT) in $(F) with goal OPTION_PERCENTAGE of 10% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

10 percent of $15 SHORT Option priced at $0.39: $0.03900000000000001
```
asset_price: 15.06
asset_change: 0.08
asset_percentage_change: 0.53
bid: 0.35
ask: 0.39
greeks: {'delta': -0.45962350266123736, 'theta': -0.014531072172559027, 'gamma': 0.4927605750510875, 'vega': 0.01037655110231925, 'implied_violatility': 30.80826997756958}
```


#### TESTING POSITION 5/5(EXIT)
 In order to exit a SHORT position (PUT) in $(F)  with goal OPTION_PERCENTAGE of 15% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

15 percent of $15 SHORT Option priced at $0.39: $0.058499999999999996
```
asset_price: 14.86
asset_change: -0.12
asset_percentage_change: -0.8
bid: 0.39
ask: 0.45
greeks: {'delta': -0.5256030566114234, 'theta': -0.023404957724889132, 'gamma': 0.3078211009767406, 'vega': 0.010270301290477114, 'implied_violatility': 50.13573169708252}
```

## Testing LONG positions


#### TESTING POSITION 1/5 (ENTRY)
 In order to enter a LONG position (CALL) in $(F) with goal ASSET_SET_PRICE of $14.88 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 14.88
asset_change: -0.1
asset_percentage_change: -0.67
bid: 0.34
ask: 0.39
greeks: {'delta': 0.46550859189932287, 'theta': -0.01798965585703377, 'gamma': 0.39913535634153535, 'vega': 0.010266814370134296, 'implied_violatility': 38.54870796203613}
```


#### TESTING POSITION 1/5(EXIT)
 In order to exit a LONG position (CALL) in $(F)  with goal ASSET_SET_PRICE of $15.09 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 15.09
asset_change: 0.11
asset_percentage_change: 0.73
bid: 0.39
ask: 0.45
greeks: {'delta': 0.5488864226924064, 'theta': -0.01817517947413471, 'gamma': 0.39207079111881016, 'vega': 0.010372228819376219, 'implied_violatility': 38.55043649673462}
```


#### TESTING POSITION 2/5 (ENTRY)
 In order to enter a LONG position (CALL) in $(F) with goal ASSET_PERCENTAGE of 0.6% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

0.6 percent of $14.98: $0.08988
```
asset_price: 14.9
asset_change: -0.08
asset_percentage_change: -0.53
bid: 0.35
ask: 0.39
greeks: {'delta': 0.4734923370546176, 'theta': -0.01804152685385441, 'gamma': 0.39921234220477636, 'vega': 0.010296417487602669, 'implied_violatility': 38.54870796203613}
```


#### TESTING POSITION 2/5(EXIT)
 In order to exit a LONG position (CALL) in $(F)  with goal ASSET_PERCENTAGE of 0.8% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

0.8 percent of $14.98: $0.11984
```
asset_price: 15.1
asset_change: 0.12
asset_percentage_change: 0.8
bid: 0.39
ask: 0.45
greeks: {'delta': 0.5528021384354803, 'theta': -0.018165108834009312, 'gamma': 0.39129719903422716, 'vega': 0.01036598484937695, 'implied_violatility': 38.552284240722656}
```


#### TESTING POSITION 3/5 (ENTRY)
 In order to enter a LONG position (CALL) in $(F) with goal OPTION_SET_PRICE of $0.35 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 14.9
asset_change: -0.08
asset_percentage_change: -0.53
bid: 0.35
ask: 0.39
greeks: {'delta': 0.4734448429119069, 'theta': -0.018025259770246344, 'gamma': 0.3995662697674073, 'vega': 0.010296335726448991, 'implied_violatility': 38.51425647735596}
```


#### TESTING POSITION 3/5(EXIT)
 In order to exit a LONG position (CALL) in $(F)  with goal OPTION_SET_PRICE of $0.45 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 15.1
asset_change: 0.12
asset_percentage_change: 0.8
bid: 0.39
ask: 0.45
greeks: {'delta': 0.5528331785074418, 'theta': -0.018143294599245127, 'gamma': 0.39175950099525014, 'vega': 0.010365876807191178, 'implied_violatility': 38.506388664245605}
```


#### TESTING POSITION 4/5 (ENTRY)
 In order to enter a LONG position (CALL) in $(F) with goal OPTION_TICK of $0.05 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 14.88
asset_change: -0.1
asset_percentage_change: -0.67
bid: 0.34
ask: 0.39
greeks: {'delta': 0.4654478004516084, 'theta': -0.017971504252642354, 'gamma': 0.3995279021917668, 'vega': 0.010266678312840484, 'implied_violatility': 38.51032257080078}
```


#### TESTING POSITION 4/5(EXIT)
 In order to exit a LONG position (CALL) in $(F)  with goal OPTION_TICK of $0.15 (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

```
asset_price: 15.25
asset_change: 0.27
asset_percentage_change: 1.8
bid: 0.39
ask: 0.54
greeks: {'delta': 0.6105873317923372, 'theta': -0.01775802362813776, 'gamma': 0.37648446394941953, 'vega': 0.010153176568382946, 'implied_violatility': 38.47825527191162}
```


#### TESTING POSITION 5/5 (ENTRY)
 In order to enter a LONG position (CALL) in $(F) with goal OPTION_PERCENTAGE of 10% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

10 percent of $15 LONG Option priced at $0.39: $0.03900000000000001
```
asset_price: 14.9
asset_change: -0.08
asset_percentage_change: -0.53
bid: 0.35
ask: 0.39
greeks: {'delta': 0.4734448429119069, 'theta': -0.018025259770246344, 'gamma': 0.3995662697674073, 'vega': 0.010296335726448991, 'implied_violatility': 38.51425647735596}
```


#### TESTING POSITION 5/5(EXIT)
 In order to exit a LONG position (CALL) in $(F)  with goal OPTION_PERCENTAGE of 15% (current ask: 0.39, current asset price: 14.98), the following price change needs to occur

15 percent of $15 LONG Option priced at $0.39: $0.058499999999999996
```
asset_price: 15.1
asset_change: 0.12
asset_percentage_change: 0.8
bid: 0.39
ask: 0.45
greeks: {'delta': 0.5528331785074418, 'theta': -0.018143294599245127, 'gamma': 0.39175950099525014, 'vega': 0.010365876807191178, 'implied_violatility': 38.506388664245605}
```


Completed testing  of (F) CALL in 0 minutes and 2 seconds


options.py
import datetime
import requests
import mibian
from functions import calculate_percentage
from enums import PositionType, PriceGoalChangeType
from polygon import OptionsClient
import json
# class MPBOptionsClient():
#     _api_endpoint = "https://api.polygon.io"
#     # def __init__(self):
#     def __init__(self,api_key, *args, **kwargs):
#         self.api_key=api_key
#         # super().__init__(*args, **kwargs)
#         # self.time = datetime.now()
#     def load_options_contracts(self, ticker, module_config):
#
#         #ok so the idea here is start at 31 days DTE and look forward
#         now = datetime.datetime.now()
#
#         for i in range(31, 90): #look 90 days out for contracts
#             url = f"{self._api_endpoint}/v3/reference/options/contracts?apiKey={self.api_key}&underlying_ticker={ticker}&expiration_date=2023-08-19&expired=false"
#             r = requests.get(url)
#             raw_data = json.loads(r.text)
#             if len(raw_data['result']) > 0:
#                 return [x['ticker'] for x in raw_data['results']]
#             print(json.loads(r.text))
#         return
#     # pass
def calculate_dte(expiration_date):
    return (datetime.datetime.strptime(expiration_date,'%Y-%m-%d')-datetime.datetime.now()).days
def calculate_price_change_for_entry(price_goal_change_type, position_type,asset_price,strike_price, price_goal, ask,  dte):
    ##in THEORY this should work for futures too?
    #basically, how much does the price need to change before our bid is met
    #can't think of a better way to do this than by the penny
    # _tmp_iv =

    if price_goal_change_type   in [PriceGoalChangeType.ASSET_POINTS, PriceGoalChangeType.ASSET_PERCENTAGE, PriceGoalChangeType.ASSET_SET_PRICE]:
        #ok so basically we need to figure out what the bid is at our target asking price
        if price_goal_change_type == PriceGoalChangeType.ASSET_POINTS:
            _asset_price = asset_price - price_goal
        elif price_goal_change_type == PriceGoalChangeType.ASSET_PERCENTAGE:
            p = float(float(price_goal / 100) * asset_price)
            print(f"{price_goal} percent of ${asset_price}: ${p}")
            _asset_price = asset_price - p*(1 if position_type == PositionType.LONG else -1)
        else:
            _asset_price = price_goal
        _tmp_price = asset_price
        _tmp_ask = ask
        if position_type == PositionType.LONG:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
        else:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], putPrice=_tmp_ask).impliedVolatility
        greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
        per_cent_delta = float(greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta )/ 100
        while round(_tmp_price,2) <= round(_asset_price,2):# and _tmp_ask > 0.05: #stop at a 0.05 ask, etrade order price min increment lol
            if round(_tmp_price, 2) == round(_asset_price, 2):
                break
            _tmp_price = _tmp_price -  0.01 *(1 if position_type == PositionType.LONG else -1) # one send decrements
            _tmp_ask = _tmp_ask - per_cent_delta * (1 if position_type == PositionType.LONG else -1)
            if position_type == PositionType.LONG:
                _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
            else:
                _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], putPrice=_tmp_ask).impliedVolatility
            greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
            per_cent_delta = float(greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta )/ 100
        bid = _tmp_ask
        # _tmp_price = _asset_price
    else:
        if position_type == PositionType.LONG:
            _tmp_iv = mibian.BS([asset_price, strike_price, 0, dte], callPrice=ask).impliedVolatility
        else:
            _tmp_iv = mibian.BS([asset_price, strike_price, 0, dte], putPrice=ask).impliedVolatility
        greeks = mibian.BS([asset_price, strike_price, 0, dte], volatility=_tmp_iv)
        per_cent_delta = float(greeks.callDelta / 100 if position_type == PositionType.LONG else greeks.putDelta / 100)

        #basically for all cases but asset changes, we can more ore less re-use the existing code, we just need to set our bid accordingly
        # if price_goal_change_type == PriceGoalChangeType.OPTION_SET_PRICE:
        if price_goal_change_type == PriceGoalChangeType.OPTION_TICK:
            bid = ask - price_goal
        elif price_goal_change_type == PriceGoalChangeType.OPTION_PERCENTAGE:
            p = float(float(price_goal/100)*ask)
            print(f"{price_goal} percent of ${strike_price} {position_type} Option priced at ${ask}: ${p}")
            bid  = ask -p
        else:
            bid = price_goal

    bid = round(bid,2)
    # per_cent_gamma = float(gamma / 100)
    #don't care about vega rn
    _tmp_ask = ask
    _tmp_price = asset_price
    # if bid > ask:
    #     raise Exception(f"Just slap the ask {ask}, it's less than your bid {bid}")

    # while _tmp_ask > bid and _tmp_ask > 0.05: #stop at a 0.05 ask, etrade order price min increment lol
    while _tmp_ask >= bid and _tmp_ask > 0.05: #stop at a 0.05 ask, etrade order price min increment lol
        if _tmp_ask <= bid:
            break
        _tmp_price = _tmp_price -  0.01 *( 1 if position_type == PositionType.LONG else -1)
        _tmp_ask = _tmp_ask - per_cent_delta *(1 if position_type == PositionType.LONG else -1)
        if position_type == PositionType.LONG:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
        else:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], putPrice=_tmp_ask).impliedVolatility
        greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
        per_cent_delta = float(greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta ) /100



    if round(bid) >= round(_tmp_ask):
        #assuming we get here, we know that the order would have been filled
        return {
            "asset_price": round(_tmp_price,2),
            "asset_change": round(float(_tmp_price - asset_price),2),
            "asset_percentage_change": round(calculate_percentage(float(_tmp_price - asset_price), asset_price),2),
            "bid": bid,
            "ask": ask,
            "greeks": {
                "delta": greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta,
                "theta": greeks.callTheta if position_type == PositionType.LONG else greeks.putTheta,
                "gamma": greeks.gamma,
                "vega": greeks.vega,
                "implied_violatility": _tmp_iv
            }
        }

    else:
        raise Exception(f"Decreased call price from {asset_price} ==> {_tmp_price} did not reach bid price")


def calculate_price_change_for_exit(price_goal_change_type, position_type,asset_price,strike_price, price_goal, ask,  dte):
    ##in THEORY this should work for futures too?
    #basically, how much does the price need to change before our bid is met
    #can't think of a better way to do this than by the penny
    # _tmp_iv =
    if position_type == PositionType.LONG:
        _tmp_iv = mibian.BS([asset_price, strike_price, 0, dte], callPrice=ask).impliedVolatility
    else:
        _tmp_iv = mibian.BS([asset_price, strike_price, 0, dte], putPrice=ask).impliedVolatility
    # _tmp_iv = mibian.BS([asset_price, strike_price, 0, dte], callPrice=ask).impliedVolatility
    greeks = mibian.BS([asset_price, strike_price, 0, dte], volatility=_tmp_iv)
    per_cent_delta = float(greeks.callDelta / 100 if position_type == PositionType.LONG else greeks.putDelta / 100)
    # _ticker_price_change_value = (0.1/100) * asset_price
    if price_goal_change_type   in [PriceGoalChangeType.ASSET_POINTS, PriceGoalChangeType.ASSET_PERCENTAGE, PriceGoalChangeType.ASSET_SET_PRICE]:
        #ok so basically we need to figure out what the bid is at our target asking price
        if price_goal_change_type == PriceGoalChangeType.ASSET_POINTS:
            _asset_price = asset_price - price_goal
        elif price_goal_change_type == PriceGoalChangeType.ASSET_PERCENTAGE:
            p = float(float(price_goal / 100) * asset_price)
            print(f"{price_goal} percent of ${asset_price}: ${p}")
            _asset_price = asset_price - p *(-1 if position_type == PositionType.LONG else 1)
        else:
            _asset_price = price_goal
        _tmp_price = asset_price
        _tmp_ask = ask
        if position_type == PositionType.LONG:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
        else:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], putPrice=_tmp_ask).impliedVolatility
        # _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
        greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
        per_cent_delta = float(greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta )/ 100
        while round(_asset_price,2) != round(_tmp_price,2):
        # while (round(_tmp_price,2) <= round(_asset_price,2) if position_type == PositionType.LONG else _asset_price <= round(_tmp_price,2)) :# and _tmp_ask > 0.05: #stop at a 0.05 ask, etrade order price min increment lol
            # if round(_tmp_price, 2) == round(_asset_price, 2):
            #     break
            _tmp_price = _tmp_price - 0.01 * (-1 if position_type == PositionType.LONG else 1)  # one send decrements
            _tmp_ask = _tmp_ask - per_cent_delta * (-1 if position_type == PositionType.LONG else 1)
            if position_type == PositionType.LONG:
                _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
            else:
                _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], putPrice=_tmp_ask).impliedVolatility
            greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
            per_cent_delta = float(greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta )/ 100
        bid = _tmp_ask
        # _tmp_price = _asset_price
    else:

        #basically for all cases but asset changes, we can more ore less re-use the existing code, we just need to set our bid accordingly
        # if price_goal_change_type == PriceGoalChangeType.OPTION_SET_PRICE:
        if price_goal_change_type == PriceGoalChangeType.OPTION_TICK:
            bid = ask + price_goal
        elif price_goal_change_type == PriceGoalChangeType.OPTION_PERCENTAGE:
            p = float(float(price_goal/100)*ask)# * (-1 if position_type == PositionType.LONG else 1 )
            print(f"{price_goal} percent of ${strike_price} {position_type} Option priced at ${ask}: ${p}")
            bid  = ask +p
            pass
            # _tmp_ask =
        else:
            bid = price_goal

    # bid = bid
    # per_cent_gamma = float(gamma / 100)
    #don't care about vega rn
    _tmp_ask = ask
    _tmp_price = asset_price
    # if bid > ask:
    #     raise Exception(f"Just slap the ask {ask}, it's less than your bid {bid}")

    # while _tmp_ask < bid: #stop at a 0.05 ask, etrade order price min increment lol
    # while (round(_tmp_ask,2) < round(bid,2) if position_type == PositionType.LONG else round(bid,2) < round(_tmp_ask,2)):
    while _tmp_ask <= bid:# and _tmp_ask > 0.05: #stop at a 0.05 ask, etrade order price min increment lol

        # print(_tmp_price)
        if _tmp_ask >= bid:
            break
        _tmp_price = _tmp_price - 0.01 * (-1 if position_type == PositionType.LONG else 1)
        _tmp_ask = _tmp_ask - per_cent_delta * (-1 if position_type == PositionType.LONG else 1)
        if position_type == PositionType.LONG:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
        else:
            _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], putPrice=_tmp_ask).impliedVolatility
        greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
        per_cent_delta = float(greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta ) /100

        # if round(_tmp_ask, 2) == round(bid, 2):
        #     break


    # else:
    #     while _tmp_ask > bid and _tmp_ask > 0.05:  # stop at a 0.05 ask, etrade order price min increment lol
    #         _tmp_price = _tmp_price + 0.01  # one send decrements
    #         _tmp_ask = _tmp_ask + per_cent_delta
    #         # ok so calculate new ask and price first, we will need them to calculate the rest
    #         #
    #         _tmp_iv = mibian.BS([_tmp_price, strike_price, 0, dte], callPrice=_tmp_ask).impliedVolatility
    #         greeks = mibian.BS([_tmp_price, strike_price, 0, dte], volatility=_tmp_iv)
    #         per_cent_delta = float(greeks.putDelta / 100)

    if round(bid,2) <= round(_tmp_ask,2):
        #assuming we get here, we know that the order would have been filled
        return {
            "asset_price": round(_tmp_price,2),
            "asset_change": round(float(_tmp_price - asset_price),2),
            "asset_percentage_change": round(calculate_percentage(float(_tmp_price - asset_price), asset_price),2),
            "bid": round(ask,2),
            "ask": round(bid,2),
            "greeks": {
                "delta": greeks.callDelta if position_type == PositionType.LONG else greeks.putDelta,
                "theta": greeks.callTheta if position_type == PositionType.LONG else greeks.putTheta,
                "gamma": greeks.gamma,
                "vega": greeks.vega,
                "implied_violatility": _tmp_iv
            }
        }

    else:
        raise Exception(f"Decreased call price from {asset_price} ==> {_tmp_price} did not reach bid price")
def calculate_price_change_percentages(stock_price, position_type, bid, ask, delta):
    pass


def load_active_contracts(ticker,client, module_config):
    pass
    #ok so here we load the active contracts for the ticker

def load_options_contracts(ticker, module_config):

    # ok so the idea here is start at 31 days DTE and look forward
    now = datetime.datetime.now()

    for i in range(31, 90):  # look 90 days out for contracts
        url = f"{module_config['api_endpoint']}/v3/reference/options/contracts?apiKey={module_config['api_key']}&underlying_ticker={ticker}&expiration_date={(now+datetime.timedelta(days=i)).strftime('%Y-%m-%d')}&expired=false"
        r = requests.get(url)
        raw_data = json.loads(r.text)
        if len(raw_data['results']) > 0:
            return [x['ticker'] for x in raw_data['results']]
        print(json.loads(r.text))
    return



def mcv_load_options_contracts(ticker, module_config):

    # ok so the idea here is start at 31 days DTE and look forward
    now = datetime.datetime.now()
    expires = {}
    for i in range(31, 90):  # look 90 days out for contracts
        url = f"{module_config['api_endpoint']}/v3/reference/options/contracts?apiKey={module_config['api_key']}&underlying_ticker={ticker}&expiration_date={(now+datetime.timedelta(days=i)).strftime('%Y-%m-%d')}&expired=false"
        r = requests.get(url)
        raw_data = json.loads(r.text)
        if len(raw_data['results']) > 0:
            expires[(now+datetime.timedelta(days=i)).strftime('%Y-%m-%d')]=[x['ticker'] for x in raw_data['results']]
        print(json.loads(r.text))
    return expires